frame/balances/src/lib.rs

```diff
@@ -608,1602 +466,611 @@ pub mod pallet {
+			Self::try_mutate_account(who, |a, _| -> Result<R, DispatchError> { Ok(f(a)) })
+		}
+
+		/// Mutate an account to some new value, or delete it entirely with `None`. Will enforce
+		/// `ExistentialDeposit` law, annulling the account as needed. This will do nothing if the
+		/// result of `f` is an `Err`.
+		///
+		/// It returns both the result from the closure, and an optional amount of dust
+		/// which should be handled once it is known that all nested mutates that could affect
+		/// storage items what the dust handler touches have completed.
+		///
+		/// NOTE: Doesn't do any preparatory work for creating a new account, so should only be used
+		/// when it is known that the account already exists.
+		///
+		/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
+		/// the caller will do this.
+		pub(crate) fn try_mutate_account<R, E: From<DispatchError>>(
+			who: &T::AccountId,
+			f: impl FnOnce(&mut AccountData<T::Balance>, bool) -> Result<R, E>,
+		) -> Result<(R, Option<T::Balance>), E> {
+			Self::ensure_upgraded(who);
+			let result = T::AccountStore::try_mutate_exists(who, |maybe_account| {
+				let is_new = maybe_account.is_none();
+				let mut account = maybe_account.take().unwrap_or_default();
+				let did_provide = account.free >= T::ExistentialDeposit::get();
+				let did_consume =
+					!is_new && (!account.reserved.is_zero() || !account.frozen.is_zero());
+
+				let result = f(&mut account, is_new)?;
+
+				let does_provide = account.free >= T::ExistentialDeposit::get();
+				let does_consume = !account.reserved.is_zero() || !account.frozen.is_zero();
+
+				if !did_provide && does_provide {
+					frame_system::Pallet::<T>::inc_providers(who);
+				}
+				if did_consume && !does_consume {
+					frame_system::Pallet::<T>::dec_consumers(who);
+				}
+				if !did_consume && does_consume {
+					frame_system::Pallet::<T>::inc_consumers(who)?;
+				}
+				if did_provide && !does_provide {
+					// This could reap the account so must go last.
+					frame_system::Pallet::<T>::dec_providers(who).map_err(|r| {
+						if did_consume && !does_consume {
+							// best-effort revert consumer change.
+							let _ = frame_system::Pallet::<T>::inc_consumers(who).defensive();
+						}
+						if !did_consume && does_consume {
+							let _ = frame_system::Pallet::<T>::dec_consumers(who);
+						}
+						r
+					})?;
+				}
 
-	// Burn funds from the total issuance, returning a positive imbalance for the amount burned.
-	// Is a no-op if amount to be burned is zero.
-	fn burn(mut amount: Self::Balance) -> Self::PositiveImbalance {
-		if amount.is_zero() {
-			return PositiveImbalance::zero()
-		}
-		<TotalIssuance<T, I>>::mutate(|issued| {
-			*issued = issued.checked_sub(&amount).unwrap_or_else(|| {
-				amount = *issued;
-				Zero::zero()
+				let maybe_endowed = if is_new { Some(account.free) } else { None };
+				let maybe_account_maybe_dust = Self::post_mutation(who, account);
```

Comment:
Might be simpler if we drop `post_mutation` abstraction. 
It's not reused, and visible only within the crate.